src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -396,32 +397,101 @@ internal Symbol BindNamespaceOrTypeOrAliasSymbol(ExpressionSyntax syntax, Diagno
 
         private TypeSymbol BindTupleType(TupleTypeSyntax syntax, DiagnosticBag diagnostics)
         {
-            var count = syntax.Elements.Count;
-            var types = ArrayBuilder<TypeSymbol>.GetInstance(count);
-            var names = ArrayBuilder<string>.GetInstance(count);
-            for (int i = 0; i < count; i++)
+            int numElements = syntax.Elements.Count;
+            var types = ArrayBuilder<TypeSymbol>.GetInstance(numElements);
+            ArrayBuilder<string> elementNames = null;
+
+            // set of names already used
+            var uniqueFieldNames = PooledHashSet<string>.GetInstance();
+            int countOfExplicitNames = 0;
+
+            for (int i = 0; i < numElements; i++)
             {
                 var argumentSyntax = syntax.Elements[i];
 
                 var argumentType = BindType(argumentSyntax.Type, diagnostics);
                 types.Add(argumentType);
 
-                // PROTOTYPE the error handling on missing, reserved or non-unique names should be fixed
-                var nameSyntax = argumentSyntax.Name;
-                var name = (object)nameSyntax != null ? nameSyntax.Identifier.ValueText : TupleTypeSymbol.UnderlyingMemberName(i);
-                names.Add(name);
+                string name = argumentSyntax.Name?.Identifier.ValueText;
+
+                // validate name if we have one
+                if (name != null)
+                {
+                    countOfExplicitNames++;
+                    CheckTupleMemberName(name, i, argumentSyntax.Name, diagnostics, uniqueFieldNames);
+                }
+                CollectTupleFieldMemberNames(name, i + 1, numElements, ref elementNames);
             }
+            uniqueFieldNames.Free();
 
-            var typesArray = types.ToImmutableAndFree();
-            if (typesArray.Length < 2 || typesArray.Length > 7)
+            if (countOfExplicitNames != 0 && countOfExplicitNames != numElements)
             {
+                Error(diagnostics, ErrorCode.ERR_TupleExplicitNamesOnAllMembersOrNone, syntax);
+            }
+
+            ImmutableArray<TypeSymbol> typesArray = types.ToImmutableAndFree();
+            if (typesArray.Length < 2)
+            {
+                // PROTOTYPE(tuples)
                 return new ExtendedErrorTypeSymbol(this.Compilation.Assembly.GlobalNamespace, LookupResultKind.NotCreatable, diagnostics.Add(ErrorCode.ERR_PrototypeNotYetImplemented, syntax.Location));
             }
 
-            var tuple = new TupleTypeSymbol(typesArray, names.ToImmutableAndFree(), syntax, this, diagnostics);
+            var tuple = new TupleTypeSymbol(
+                typesArray,
+                elementNames == null ?
+                    default(ImmutableArray<string>) :
+                    elementNames.ToImmutableAndFree(),
```

Comment:
When we bind tuples (declaration and creation), we currently place a name for every element whether or not the user put one. So currently they always match. 
I will add an assert since we expect them to be of equal lengths at this point.

That said, we discussed maybe not populating Item1 through 7 if the user didn't provide names (and let the underlying type handling them). 
I'm not sure I favor that, but if we did that, then we could end up with fewer element names than types.